src/abseil-cpp/preprocessed_builds.yaml.gen.py

```diff
@@ -63,38 +64,101 @@ def parse_rule(elem, package):
       testonly=get_elem_value(elem, "testonly") or False)
 
 
-def read_build(package):
+def read_bazel_build(package):
   """Runs bazel query on given package file and returns all cc rules."""
   result = subprocess.check_output(
       ["bazel", "query", package + ":all", "--output", "xml"])
   root = ET.fromstring(result)
   return [
-      parse_rule(elem, package)
+      parse_bazel_rule(elem, package)
       for elem in root
       if elem.tag == "rule" and elem.attrib["class"].startswith("cc_")
   ]
 
 
-def collect_rules(root_path):
-  """Collects and returns all rules from root path recursively."""
+def collect_bazel_rules(root_path):
+  """Collects and returns all bazel rules from root path recursively."""
   rules = []
   for cur, _, _ in os.walk(root_path):
     build_path = os.path.join(cur, "BUILD.bazel")
     if os.path.exists(build_path):
-      rules.extend(read_build("//" + cur))
+      rules.extend(read_bazel_build("//" + cur))
   return rules
 
 
+def parse_cmake_rule(str, package):
+  """Returns a rule from absl cmake rule.
+     Reference: https://github.com/abseil/abseil-cpp/blob/master/CMake/AbseilHelpers.cmake
+  """
+  kv = {}
+  bucket = None
+  lines = str.splitlines()
+  for line in lines[1:-1]:
+    if re.match("[A-Z]+", line.strip()):
+      bucket = kv.setdefault(line.strip(), [])
+    else:
+      if bucket is not None:
+        bucket.append(line.strip())
+      else:
+        assert "Illegal syntax: " + str
+  return Rule(
+      type=lines[0].rstrip("("),
+      name="absl::" + kv["NAME"][0],
+      package=package,
+      srcs=[package + "/" + f.strip('"') for f in kv.get("SRCS", [])],
+      hdrs=[package + "/" + f.strip('"') for f in kv.get("HDRS", [])],
+      textual_hdrs=[],
+      deps=kv.get("DEPS", []),
+      visibility="PUBLIC" in kv,
+      testonly="TESTONLY" in kv,
+  )
+
+
+def read_cmake_build(build_path, package):
+  """Parses given CMakeLists.txt file and returns all cc rules."""
+  rules = []
+  with open(build_path, "r") as f:
+    src = f.read()
+    for mo in re.finditer("^absl_cc_.*\(", src, re.MULTILINE):
+      mo2 = re.search("^\)", src[mo.start(0):], re.MULTILINE)
+      expr = src[mo.start(0):mo.start(0) + mo2.start(0) + 1]
+      rules.append(parse_cmake_rule(expr, package))
+  return rules
+
+
+def collect_cmake_rules(root_path):
+  """Collects and returns all cmake rules from root path recursively."""
+  rules = []
+  for cur, _, _ in os.walk(root_path):
+    build_path = os.path.join(cur, "CMakeLists.txt")
+    if os.path.exists(build_path):
+      rules.extend(read_cmake_build(build_path, cur))
+  return rules
+
+
+def pairing_bazel_and_cmake_rules(bazel_rules, cmake_rules):
+  """Returns a pair map between bazel rules and cmake rules based on
+     the similarity of the file list in the rule."""
+  pair_map = {}
+  for rule in bazel_rules:
+    best_crule, best_s = None, 0
+    for crule in cmake_rules:
```

Comment:
Looks like this is O(n^2) (assuming there are roughly as many bazel rules as cmake rules). Is that going to be a maintainability problem? How long does this script currently take to run?

Reply:
It'd run in 1 sec and it would be fine even with thousands of rules. :)
FYI, running bazel over BUILD files is slowest and it'd take several seconds.